src/Security/Authentication/Core/src/AuthenticationBuilder.cs

```diff
@@ -47,7 +47,8 @@ public AuthenticationBuilder(IServiceCollection services)
 
             if (configureOptions != null)
             {
-                optionsBuilder.Configure(configureOptions);
+                Services.AddSingleton<IConfigureOptions<TOptions>>(sp =>
```

Comment:
I think we clearly have a test hole where we aren't trying this with different service lifetimes, can you add some tests that verify that things work as expected when TService is registered as transient, scoped, and singleton.  I'm guessing this might throw if TService is scoped

Reply:
> can you add some tests that verify that things work as expected when TService is registered as transient, scoped, and singleton.

What should be the scenario of the tests? Should I create a ioc container scope in the test? or should I test it in a funtional test?

> I'm guessing this might throw if TService is scoped

yeah, this could be an issue if there's no scope a the time of reading the config.

Reply:
What do you think about using a new class instead of `ConfigureNamedOption` which doesn't need to hold to the dependency and creates it when `Configure` is called?

```csharp
internal class LazyConfigureNamedOptions<TOptions, TDep> : IConfigureNamedOptions<TOptions>, IConfigureOptions<TOptions>
        where TOptions : class
        where TDep : class
    {
        private readonly string _name;
        private readonly IServiceProvider _serviceProvider;
        private readonly Action<TOptions, TDep> _action;


        public LazyConfigureNamedOptions(string name, IServiceProvider serviceProvider, Action<TOptions, TDep> action)
        {
            _name = name;
            _serviceProvider = serviceProvider;
            _action = action;
        }

        public virtual void Configure(string name, TOptions options)
        {
            if (options == null)
            {
                throw new ArgumentNullException(nameof(options));
            }

            // Null name is used to configure all named options.
            if (_name == null || name == _name)
            {
                _action?.Invoke(options, _serviceProvider.GetService<TDep>());
            }
        }

        public void Configure(TOptions options) => Configure(Options.DefaultName, options);
    }
```

Reply:
Wouldn't this just effectively do the same thing as what the options builder configure which registers as transient?  Which is it will resolve a new instance of the service from the SP every time?

Reply:
RE test, you should have a test for each permutation of the dep, Singleton, Transient and Scoped, and they exercise that the options are configured with the correct service instances per registered lifetime, singleton should always get the same instance, transient always different, scoped we would expect the instance is the same within a scope and outside.  You don't need to do this in a functional test, you should be able to just create a service collection and provider and scopes as needed and directly exercise the auth options, I think...

Reply:
You might have to mock up some stuff for auth

Reply:
> Wouldn't this just effectively do the same thing as what the options builder configure which registers as transient? Which is it will resolve a new instance of the service from the SP every time?

This got me thinking... what was the purpose of this PR again 😁 ?

Before this https://github.com/dotnet/aspnetcore/commit/1f76cce14ac4e4698a554b65a24f28000b50396e#r40670134, there were no dependency involved and a singleton `IConfigureOptions` was being registered, which only depends on `Action<TOptions>`

```
services.AddSingleton<IConfigureOptions<TOptions>>(new ConfigureNamedOptions<TOptions>(name, 
configureOptions));
```
and then it allows to have a dependency and has changed to 

```
Services.AddTransient<IConfigureOptions<TOptions>>(sp => 
    new ConfigureNamedOptions<TOptions, TDep>(Name, sp.GetRequiredService<TDep>(), configureOptions));
```

I know that the lifetime of `IConfigureOptions<TOptions>` has changed, but maybe I'm missing something, I don't know what difference it makes? how does this change could break the current applications?

Reply:
Its hard to say, but to make up an example, what if someone wrote some code that cached IConfigureOptions instances of certain auth types expecting that they were singletons so that was safe, and now they are all of a sudden getting different instances every time

Reply:
> singleton should always get the same instance, transient always different, scoped we would expect the instance is the same within a scope and different once outside.

wait a second ... , now that a singleton `IConfigureOptions` is registered and it wants an instance of dependency in its ctor, no matter which lifetime the dependency has, it will always be the same instance. Unless if we go with [LazyConfigureNamedOptions](https://github.com/dotnet/aspnetcore/pull/24052#discussion_r456664407) which the `IConfigureOptions` itself is singleton but the dependency is being resolved each time `Configure` is called and the lifetime in handled by service provider.

> Its hard to say, but to make up an example, what if someone wrote some code that cached IConfigureOptions instances of certain auth types expecting that they were singletons so that was safe, and now they are all of a sudden getting different instances every time

It is possible to just handle these cases with an announcement? :sweat_smile:

Reply:
Right, the old behavior with the auth schemes is for the options to be singletons and hence not change.  That's why this was a breaking change.

The OptionsBuilder apis behave differently, its possible you need to implement this change differently for it to work, i.e. leave the AddSchemeHelper alone, and call a different codepath that uses the transient options builder api when resolving services.